tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py

```diff
@@ -430,74 +524,334 @@ def apply_gradients(grads_and_vars, global_step=None, name=None):
+        dense_grads_and_vars_aggregated_out = []
+        sparse_grads_and_vars_unaggregated_out = []
+        test_unaggregated_lambda = lambda x: isinstance(
+            x[1], DistributedVariableWrapper)
+        for g_and_v in grads_and_vars_in:
+          if test_unaggregated_lambda(g_and_v):
+            sparse_grads_and_vars_unaggregated_out.append(g_and_v)
+          else:
+            dense_grads_and_vars_aggregated_out.append(g_and_v)
+        return tuple(dense_grads_and_vars_aggregated_out), tuple(
+            sparse_grads_and_vars_unaggregated_out)
+
+      dense_grads_and_vars_aggregated, sparse_grads_and_vars_unaggregated = DistributedVariableWrapperFilter(
+          grads_and_vars)
+      dense_grads_and_vars_aggregated = self.aggregate_gradients(
+          dense_grads_and_vars_aggregated)
+      grads_and_vars = dense_grads_and_vars_aggregated
+      for g_and_v in sparse_grads_and_vars_unaggregated:
+        grads_and_vars.append(g_and_v)
+
+    return super(keras_OptimizerV2, self).apply_gradients(grads_and_vars,
+                                                          name=name)
+
+  def compute_gradients_horovod_wrapper(compute_gradients_horovod_func):
+
+    def compute_gradients_horovod_wrapper_impl(*args, **kwargs):
+      if self._finish_register_tw is False:
+        var_list = kwargs.get('var_list', None)
+        if var_list is None:
+          raise Exception(
+              "var_list parameter must not be None when using compute_gradients function!"
+          )
+        _, trainable_wrapper_vars = trainable_wrapper_filter(var_list)
+        for tw in trainable_wrapper_vars:
+          self.register_local_var(tw)
+        self._finish_register_tw = True
+      return compute_gradients_horovod_func(*args, **kwargs)
+
+    return compute_gradients_horovod_wrapper_impl
+
   if isinstance(self, optimizer.Optimizer):
+    # TF1 Optimizer
     self._get_or_make_slot = _get_or_make_slot
     self._get_or_make_slot_with_initializer = _get_or_make_slot_with_initializer
     self._zeros_slot = _zeros_slot
-    if self._hvd_sync:
-      self.apply_gradients = apply_gradients
-  elif isinstance(self, optimizer_v2.OptimizerV2) or isinstance(
-      self, keras_optimizer):
-    if self._hvd_sync:
+    # TODO(MoFHeka): Support distribute strategy with tf.compat.v1.train.Optimizer
+    # self.get_slot = get_slot_v1
+    if self._custom_sync:
+      self.apply_gradients = apply_gradients_sync_v1
+    elif 'horovod' in str(self):
+      # If use Horovod keras optimizer wrapper
+      if not hasattr(self, 'register_local_var'):
+        import horovod
+        raise Exception(
+            f"Input optimizer must has register_local_var function! Your Horovod version is {horovod.__version__}, please upgrade to greater than 0.26.0"
+        )
+      self._finish_register_tw = False
+      compute_gradients_horovod_v1 = self.compute_gradients
+      self.compute_gradients = compute_gradients_horovod_wrapper(
```

Comment:
This code may be hard to be accepted by the TensorFlow core. Do we have any plan or scheme about how to deal with this part of the code when moving the trainable wrapper to the TensorFlow core?

Reply:
Remove all code related to horovod, and remove all code about low version compatibility?